### PR TITLE
EHN: export to dataframe and CSV

### PR DESCRIPTION
**Pull request by glemaitre**
TODO:

- [x] Decide if we do automatic filename export base on the meta data.
- [x] Add an example to show how to open some spectrum, export and repopen. We need to merge #47 before thought.

### CODE REVIEW

File: specio/core/util.py [hunk 1]
```diff
@@ -101,6 +103,39 @@ def __init__(self, amplitudes, wavelength, meta=None):
             self._validate_amplitudes_wavelength(amplitudes, wavelength)
         self.meta = meta if meta is not None else {}
 
+    def to_dataframe(self):
+        """Export to the Spectrum to a pandas DataFrame.
```
Comment (jorisvandenbossche) [inline specio/core/util.py]: export ~~to~~ the spectrum

File: specio/core/util.py [hunk 2]
```diff
@@ -101,6 +103,39 @@ def __init__(self, amplitudes, wavelength, meta=None):
...
+
+        Returns
+        -------
+        df_spectrum : DataFrame,
+            A pandas DataFrame containing the information from a Spectrum.
```
Comment (jorisvandenbossche) [inline specio/core/util.py]: maybe say something like "with the index the filename(s) and the columns the wavelengths"

File: specio/core/util.py [hunk 3]
```diff
@@ -101,6 +103,39 @@ def __init__(self, amplitudes, wavelength, meta=None):
...
+        ----------
+        filename : str
+            File path
+
+        Returns
+        -------
+        None
+
+        """
+        self.to_dataframe().to_csv(filename)
```
Comment (jorisvandenbossche) [inline specio/core/util.py]: do we want to pass `**kwargs` to to_csv? (not sure if there are that would be needed though, as that just opens the possibility to produce less standard csv files)

File: specio/core/tests/test_util.py [hunk 4]
```diff
@@ -4,11 +4,21 @@
...
 # License: BSD 3 clause
 
+import os
+from tempfile import mkdtemp
+from shutil import rmtree
+
 import pytest
 
 import numpy as np
+import pandas as pd
```
Comment (jorisvandenbossche) [inline specio/core/tests/test_util.py]: *in principle* you can import this with `pd = pytest.importorskip('pandas')` if you also want to keep pandas as optional dependency for testing. But personally I think that for pandas this is less of a problem

File: specio/core/tests/test_util.py [hunk 5]
```diff
@@ -34,6 +44,30 @@ def test_spectrum(spectrum, wavelength, metadata):
     assert spec.meta == {'kind': 'random'}
 
 
+def test_spectrum_to_dataframe():
+    spec = specread(load_spc_path())
+    df_spec = spec.to_dataframe()
+    assert_array_equal(df_spec.index.values, np.array(spec.meta['filename']))
+    assert_allclose(df_spec.columns.values, spec.wavelength)
+    assert_allclose(df_spec.values, np.atleast_2d(spec.amplitudes))
```
Comment (jorisvandenbossche) [inline specio/core/tests/test_util.py]: This only tests the case with a Spectrum read from a single file?

File: specio/core/tests/test_util.py [hunk 6]
```diff
@@ -34,6 +44,30 @@ def test_spectrum(spectrum, wavelength, metadata):
...
+
+    try:
+        spec.to_csv(filename)
+        df = pd.read_csv(filename, index_col=0)
+        df.columns = df.columns.astype(float)
+        assert_frame_equal(df, spec.to_dataframe(),
+                           check_exact=False,
+                           check_less_precise=True)
+    finally:
+        rmtree(tmp_dir)
```
Comment (jorisvandenbossche) [inline specio/core/tests/test_util.py]: You can also use this the tempfile library in some way as a context manager I think (which could be cleaner)

Comment (jorisvandenbossche) [reply specio/core/tests/test_util.py]: Or maybe even neater, a builtin pytest fixture :-) 
(that's actually what I ended up using last week when I was searching for a way to use tempfile as a context manager to automatically do this cleaning up, and then I saw that pytest has it builtin, see eg https://github.com/geopandas/geopandas/commit/36c8bcda34e4f46315eda812b11a0d9db32d01f2#diff-318778081cf3317568072c6793e5e95aR76)

**Comment by jorisvandenbossche** [comment]
One api question though: are we sure we want it in this shape, and not transposed? (wavelengths as index, different files as columns)
I am trying to remember what we typically needed for the analysis, but eg for plotting I think we typically needed to do a `.T` ?

**Comment by jorisvandenbossche** [comment]
> Decide if we do automatic filename export base on the meta data.

What do you mean with this?

File: specio/core/tests/test_util.py (hunk 5 above)
Comment (glemaitre) [reply specio/core/tests/test_util.py]: yes for the moment. I need to add some data to check that this is working as expected (I already tested on another set indeed).

File: specio/core/util.py (hunk 3 above)
Comment (glemaitre) [reply specio/core/util.py]: I would not because my CSV reader will be not valid afterwords

**Comment by jorisvandenbossche** [comment]
> One api question though: are we sure we want it in this shape, and not transposed? (wavelengths as index, different files as columns)

Although now it is consistent with how the amplitudes array is stored of course ..

**Comment by glemaitre** [comment]
> What do you mean with this?

I could think about:

```python
spec.to_csv()
```

without giving any name. It will trigger a default naming depending of the meta information. Like this we could think that the command line do not even required an output

**Comment by glemaitre** [comment]
> One api question though: are we sure we want it in this shape, and not transposed? (wavelengths as index, different files as columns)

Only issue is the plotting using `.T`. In scikit-learn for instance the shape will be fine.

File: specio/core/util.py (hunk 3 above)
Comment (jorisvandenbossche) [reply specio/core/util.py]: that's a good reason :)

**Comment by jorisvandenbossche** [comment]
> Only issue is the plotting using .T. In scikit-learn for instance the shape will be fine.

Yes, that is an additional reason to keep it as is.

**Pull request merged**